[Elastic] Skip store barrier and store get in host assign

fduwjj opened the pull request:
Summary: Sometimes, when the worker and role are the same, users want to skip TCPStore in `_assign_worker_ranks` and barrier in rendezvous

Test Plan: unit test

Differential Revision: D63351662


cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @wz337 @wconstab @d4l3k @c-p-i-o

d4l3k reviewed (approved):
LGTM

d4l3k commented on test/distributed/elastic/agent/server/test/api_test.py:
```diff
@@ -542,6 +546,58 @@ def f(info) -> List[Worker]:
         )
 
+    def test_assign_worker_ranks_indentical(self):
+        os.environ["TORCH_SKIP_STORE_BARRIER"] = "1"
```
you should probably reset this at the end of the test

Commit:
[Elastic] Skip store barrier and store get in host assign (#136579)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/136579

Sometimes, when the worker and role are the same, users want to skip TCPStore in `_assign_worker_ranks` and barrier in rendezvous

Test Plan: unit test

Reviewed By: d4l3k

Differential Revision: D63351662

fduwjj commented:
reland in https://github.com/pytorch/pytorch/pull/136865
